### PR TITLE
Added relation detail analysis.

### PR DESCRIPTION
**Pull request by azreika**
Used to cache relation->clause and name->relation mappings in a program. Speeds up semantic checker substantially in some cases.

### CODE REVIEW

File: src/PrecedenceGraph.h [hunk 1]
```diff
@@ -109,6 +109,32 @@ class RecursiveClauses : public AstAnalysis {
...
+    static constexpr const char* name = "relation-detail";
+
+    RelationDetail() : AstAnalysis(name) {}
+
+    void run(const AstTranslationUnit& translationUnit) override;
+
+    void print(std::ostream& os) const override;
+
+    AstRelation* getRelation(const AstQualifiedName& name) const {
+        return nameToRelation.at(name);
```
Comment (mmcgr) [inline src/PrecedenceGraph.h]: This is throwing errors in semantic/agg_checks2 as a name is not found. The name is for an undefined relation.

**Comment by mmcgr** [comment]
Jenkins, test this please

File: src/PrecedenceGraph.h [hunk 2]
```diff
@@ -109,6 +109,43 @@ class RecursiveClauses : public AstAnalysis {
     bool computeIsRecursive(const AstClause& clause, const AstTranslationUnit& translationUnit) const;
 };
 
+/**
+ * Analysis pass mapping identifiers with relations and clauses.
+ */
+class RelationDetail : public AstAnalysis {
```
Comment (mmcgr) [inline src/PrecedenceGraph.h]: The name doesn't seem to reflect its purpose. `RelationDetailCache`, perhaps, or `RelationInfoCache`?

Comment (azreika) [reply src/PrecedenceGraph.h]: Good point, I'll change it to `RelationDetailCache`.

**Pull request merged**